### PR TITLE
feat(api): quicer:peercert/1

### LINKED ISSUE #194: Getting the peer TLS certificate
Hello, I may have missed something. Is there a way to get the peer TLS certificate, and if so how? Thank you.

### PR DESCRIPTION
**Pull request by qzhuyan**
fix #194

### CODE REVIEW

File: c_src/quicer_connection.c [hunk 1]
```diff
@@ -92,6 +92,39 @@ static QUIC_STATUS handle_connection_event_resumption_ticket_received(
...
+  else if (enif_get_resource(env, ctx, ctx_connection_t, &q_ctx))
+    {
+      c_ctx = (QuicerConnCTX *)q_ctx;
+    }
+  else
+    {
+      return ERROR_TUPLE_2(ATOM_BADARG);
+    }
+
+  assert(c_ctx);
```
Comment (qzhuyan) [inline c_src/quicer_connection.c:115]: maybe null?

File: c_src/quicer_connection.c [hunk 2]
```diff
@@ -92,6 +92,39 @@ static QUIC_STATUS handle_connection_event_resumption_ticket_received(
...
+
+  if (!c_ctx->peer_cert)
+    {
+      return SUCCESS(ATOM_UNDEFINED);
```
Comment (qzhuyan) [inline c_src/quicer_connection.c:119]: check OTP return val?

Comment (qzhuyan) [reply c_src/quicer_connection.c:119]: should be `{error,no_peercert}`

File: c_src/quicer_connection.c [hunk 3]
```diff
@@ -92,6 +92,39 @@ static QUIC_STATUS handle_connection_event_resumption_ticket_received(
...
+    }
+
+  unsigned char *data
+      = enif_make_new_binary(env, i2d_X509(c_ctx->peer_cert, NULL), &DerCert);
```
Comment (qzhuyan) [inline c_src/quicer_connection.c:123]: safer to check ret val of `i2d_X509`

File: c_src/quicer_connection.c [hunk 4]
```diff
@@ -1497,11 +1530,18 @@ static QUIC_STATUS
...
-  // @TODO peer_certificate_received
   // Only with QUIC_CREDENTIAL_FLAG_INDICATE_CERTIFICATE_RECEIVED set
   assert(QUIC_CONNECTION_EVENT_PEER_CERTIFICATE_RECEIVED == Event->Type);
   // Validate against CA certificates using OpenSSL API:s
   X509 *cert = (X509 *)Event->PEER_CERTIFICATE_RECEIVED.Certificate;
+
+  // Preserve cert in ctx
+  if (c_ctx->peer_cert)
+    {
+      X509_free(c_ctx->peer_cert);
```
Comment (qzhuyan) [inline c_src/quicer_connection.c:1541]: wrap  alloc/free to make them trackable.

File: c_src/quicer_connection.c [hunk 5]
```diff
@@ -92,6 +92,39 @@ static QUIC_STATUS handle_connection_event_resumption_ticket_received(
...
+  assert(c_ctx);
+
+  if (!c_ctx->peer_cert)
+    {
+      return ERROR_TUPLE_2(ATOM_NO_PEERCERT);
+    }
+
+  unsigned char *data
+      = enif_make_new_binary(env, i2d_X509(c_ctx->peer_cert, NULL), &DerCert);
+  i2d_X509(c_ctx->peer_cert, &data);
```
Comment (qzhuyan) [inline c_src/quicer_connection.c:124]: need a tmp var as desc in https://linux.die.net/man/3/i2d_x509

**Review by thalesmg — APPROVED** [review]

**Pull request merged**